pystiche_papers/sanakoyeu_et_al_2018/_nst.py

```diff
@@ -210,3 +223,137 @@ def optim_loop(transformer: nn.Module) -> nn.Module:
             transformer_lr_scheduler.step(epoch)
 
     return transformer
+
+
+def training(
+    content_image_loader: Union[DataLoader, str],
+    style_image_loader: Union[DataLoader, str],
+    style: Optional[str] = None,
+    impl_params: bool = True,
+) -> nn.Module:
+    r"""Training a transformer for the NST.
+
+    Args:
+        content_image_loader: Content images used as input for the ``transformer`` or
+            string to use a predefined image_loader.
+        style_image_loader: Style images used as input for the ``discriminator`` or
+            string to use a predefined image_loader.
+        style: Optional style string which is needed if the 'style_image_loader' is a
+            string.
+        impl_params: If ``True``, uses the parameters used in the reference
+            implementation of the original authors rather than what is described in
+            the paper.
+
+    """
+    device = misc.get_device()
+
+    if isinstance(content_image_loader, str):
+        root = content_image_loader
+        dataset = content_dataset(root, impl_params=impl_params)
+        content_image_loader = image_loader(dataset, impl_params=impl_params)
+
+    if isinstance(style_image_loader, str):
+        if style is not None:
+            root = style_image_loader
+            dataset = style_dataset(root, style=style, impl_params=impl_params)
+            style_image_loader = image_loader(dataset, impl_params=impl_params)
+        else:
+            raise ValueError(
+                "The style is currently '"
+                + str(style)
+                + "', please use a valid string for the style because it is needed for "
+                + "the style_dataset."
+            )
```

Comment:
1. You should use `f-strings` since they are more readable (and faster, but that is not an issue here). This would also remove the need to concatenate the strings with addition.
2. The message is to explicit. 

```suggestion
            raise ValueError(
                "The parameter 'style' cannot be omitted if 'style_image_dataset' is passed as dataset root."
            )
```